data-serving/data-service/src/controllers/case.ts

```diff
@@ -222,50 +222,55 @@ export class CasesController {
      * Perform geocoding for each case (of multiple `cases` specified in the
      * request body), in accordance with the above geocoding logic.
      *
-     * TODO: Use a batch geocode API like https://docs.mapbox.com/api/search/#batch-geocoding.
-     *
-     * @returns {object} Detailing the nature of any issues encountered.
+     * TODO: https://github.com/globaldothealth/list/issues/1131 rate limit.
      */
     batchGeocode = async (
         req: Request,
         res: Response,
         next: NextFunction,
     ): Promise<void> => {
-        const caseCount = req.body.cases.length;
         const geocodeErrors: { index: number; message: string }[] = [];
-        for (let index = 0; index < caseCount; index++) {
-            const c = req.body.cases[index];
-            try {
-                const geocodeResult = await this.geocode({
-                    body: c,
+        Promise.all(
```

Comment:
Did you test this with a large number of cases in the request body?

I'm debating changing the `batchValidate` method for doing this. Storing a Promise<N> takes _way_ more memory than storing an N, which is why that method OOMs so easily.

Reply:
Let me do some time analysis and get back to you but how is batchValidate going to help here? I'm not sure I follow.

Reply:
On the client side it's substantially faster.

Fresh run on main:

```
Sending 250 cases, total so far: 0
	Current speed: 23 cases/sec
	Max estimate: 20628 in the remaining 886 sec
Sending 250 cases, total so far: 250
	Current speed: 23 cases/sec
	Max estimate: 20625 in the remaining 875 sec
Sending 250 cases, total so far: 500
	Current speed: 20 cases/sec
	Max estimate: 17950 in the remaining 860 sec
Sending 250 cases, total so far: 750
	Current speed: 20 cases/sec
	Max estimate: 17940 in the remaining 847 sec
Sending 250 cases, total so far: 1000
	Current speed: 20 cases/sec
	Max estimate: 17970 in the remaining 836 sec
Sending 250 cases, total so far: 1250
	Current speed: 18 cases/sec
	Max estimate: 16224 in the remaining 818 sec
Sending 250 cases, total so far: 1500
	Current speed: 19 cases/sec
	Max estimate: 17083 in the remaining 807 sec
Sending 250 cases, total so far: 1750
	Current speed: 20 cases/sec
	Max estimate: 17960 in the remaining 798 sec
Sending 250 cases, total so far: 2000
	Current speed: 21 cases/sec
	Max estimate: 18924 in the remaining 794 sec
Sending 250 cases, total so far: 2250
	Current speed: 22 cases/sec
	Max estimate: 19770 in the remaining 785 sec
Sending 250 cases, total so far: 2500
	Current speed: 23 cases/sec
	Max estimate: 20644 in the remaining 778 sec
Sending 250 cases, total so far: 2750
	Current speed: 24 cases/sec
	Max estimate: 21528 in the remaining 772 sec
Sending 250 cases, total so far: 3000
	Current speed: 25 cases/sec
	Max estimate: 22450 in the remaining 768 sec
Sending 196 cases, total so far: 3250
	Current speed: 26 cases/sec
	Max estimate: 23362 in the remaining 766 sec
sent 3446 cases in 130.53173327445984 seconds
Updating upload via http://localhost:3001/api/sources/5f64c9e7f7add200286d9af5/uploads/5f64ca4cf7add200286d9af9
END RequestId: 1354de99-439a-1b8c-d115-3802e90b6760
REPORT RequestId: 1354de99-439a-1b8c-d115-3802e90b6760	Init Duration: 735.32 ms	Duration: 132701.62 ms	Billed Duration: 132800 ms	Memory Size: 128 MB	Max Memory Used: 54 MB

{"count_created":3446,"count_updated":0}
```


Run from this PR (fresh db and restarted server as well):

```
Sending 250 cases, total so far: 0
	Current speed: 14 cases/sec
	Max estimate: 12556 in the remaining 879 sec
Sending 250 cases, total so far: 250
	Current speed: 20 cases/sec
	Max estimate: 17940 in the remaining 872 sec
Sending 250 cases, total so far: 500
	Current speed: 23 cases/sec
	Max estimate: 20645 in the remaining 865 sec
Sending 250 cases, total so far: 750
	Current speed: 26 cases/sec
	Max estimate: 23308 in the remaining 858 sec
Sending 250 cases, total so far: 1000
	Current speed: 28 cases/sec
	Max estimate: 25134 in the remaining 853 sec
Sending 250 cases, total so far: 1250
	Current speed: 31 cases/sec
	Max estimate: 27788 in the remaining 848 sec
Sending 250 cases, total so far: 1500
	Current speed: 33 cases/sec
	Max estimate: 29569 in the remaining 843 sec
Sending 250 cases, total so far: 1750
	Current speed: 34 cases/sec
	Max estimate: 30526 in the remaining 839 sec
Sending 250 cases, total so far: 2000
	Current speed: 37 cases/sec
	Max estimate: 33182 in the remaining 836 sec
Sending 250 cases, total so far: 2250
	Current speed: 38 cases/sec
	Max estimate: 34116 in the remaining 832 sec
Sending 250 cases, total so far: 2500
	Current speed: 40 cases/sec
	Max estimate: 35830 in the remaining 827 sec
Sending 250 cases, total so far: 2750
	Current speed: 41 cases/sec
	Max estimate: 36784 in the remaining 824 sec
Sending 250 cases, total so far: 3000
	Current speed: 43 cases/sec
	Max estimate: 38596 in the remaining 822 sec
Sending 196 cases, total so far: 3250
	Current speed: 45 cases/sec
	Max estimate: 40346 in the remaining 820 sec
sent 3446 cases in 76.32758855819702 seconds
Updating upload via http://localhost:3001/api/sources/5f64c9e7f7add200286d9af5/uploads/5f64cdedb90c2400286dfbcd
END RequestId: a39ace62-dc83-1969-769c-e47944fe0e12
REPORT RequestId: a39ace62-dc83-1969-769c-e47944fe0e12	Init Duration: 1280.48 ms	Duration: 78621.33 ms	Billed Duration: 78700 ms	Memory Size: 128 MB	Max Memory Used: 54 MB

{"count_created":3446,"count_updated":0}
```

I haven't looked at memory usage though, we did boost the memory limits by a fair amount recently though.
I'll try to look for memory usage of local docker images and ways to graph those...

Reply:
Ah nice, that looks great.

`batchValidate` isn't going to impact this method -- I meant that I found the OOMs in `batchValidate` were largely caused by converting the request body cases to Promises (and waiting with Promises.all). Since, as mentioned, storing a Promise<N> takes way more memory than storing an N. This was mostly a problem with bulk upload, though, I think.

That said, I'm cool to resolve, since I doubt ADI -- which batches cases into 250 -- would cause an OOM.